lib-fastscrollrecyclerview/src/main/java/com/jaredrummler/fastscrollrecyclerview/FastScrollRecyclerView.java

```diff
@@ -387,8 +390,58 @@ protected void getCurScrollState(ScrollPositionState stateOut) {
     if (getLayoutManager() instanceof GridLayoutManager) {
       stateOut.rowIndex = stateOut.rowIndex / ((GridLayoutManager) getLayoutManager()).getSpanCount();
     }
-    stateOut.rowTopOffset = getLayoutManager().getDecoratedTop(child);
-    stateOut.rowHeight = child.getHeight();
+    stateOut.rowTopOffset = getLayoutManager().getDecoratedTop(child) / (float) child.getHeight();
+    stateOut.rowHeight = calculateRowHeight(child.getHeight());
+  }
+
+  /**
+   * Calculates the row height based on the average of the visible children, to handle scrolling
+   * through children with different heights gracefully
+   */
+  protected int calculateRowHeight(int fallbackHeight) {
+    LayoutManager layoutManager = getLayoutManager();
+
+    if (layoutManager instanceof LinearLayoutManager || layoutManager instanceof GridLayoutManager) {
```

Comment:
GridLayoutManager extends from LinearLayoutManager so the second condition is always true.